File: app/src/main/java/ie/dublinbuspal/android/view/search/SearchAdapter.java
package ie.dublinbuspal.android.view.search;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.Collections;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import ie.dublinbuspal.android.R;
import ie.dublinbuspal.model.route.Route;
import ie.dublinbuspal.model.stop.Stop;
import ie.dublinbuspal.util.CollectionUtils;
import ie.dublinbuspal.util.StringUtils;

public class SearchAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TITLE = 0;
    private static final int ROUTE = 1;
    private static final int BUS_STOP = 2;

    private List<Object> searchResults = Collections.emptyList();
    private final SearchQueryView view;

    SearchAdapter(SearchQueryView view) {
        this.view = view;
    }

    @Override
    @NonNull
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case TITLE:
                View titleView = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.list_item_search_title, parent, false);
                return new TitleViewHolder(titleView);
            case ROUTE:
                View routeView = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.list_item_route, parent, false);
                RouteViewHolder routeViewHolder = new RouteViewHolder(routeView);
                routeView.setOnClickListener(routeViewHolder);
                return routeViewHolder;
            case BUS_STOP:
                View busStopView = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.list_item_bus_stop, parent, false);
                BusStopViewHolder busStopViewHolder = new BusStopViewHolder(busStopView);
                busStopView.setOnClickListener(busStopViewHolder);
                return busStopViewHolder;
            default:
                throw new IllegalStateException(String.format(Locale.UK,
                        "Incorrect view type: %s", String.valueOf(viewType)));
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (position > -1 && position < searchResults.size()) {
            Object searchResult = searchResults.get(position);
            if (searchResult instanceof String && holder instanceof TitleViewHolder) {
                String text = (String) searchResult;
                TitleViewHolder titleViewHolder = (TitleViewHolder) holder;
                titleViewHolder.bind(text);
            } else if (searchResult instanceof Route && holder instanceof RouteViewHolder) {
                Route route = (Route) searchResult;
                RouteViewHolder routeViewHolder = (RouteViewHolder) holder;
                routeViewHolder.bind(route);
            } else if (searchResult instanceof Stop && holder instanceof BusStopViewHolder) {
                Stop busStop = (Stop) searchResult;
                BusStopViewHolder busStopViewHolder = (BusStopViewHolder) holder;
                busStopViewHolder.bind(busStop);
            }
        }
    }

    @Override
    public int getItemCount() {
        return searchResults.size();
    }

    @Override
    public int getItemViewType(int position) {
        Object searchResult = searchResults.get(position);
        if (searchResult instanceof String) {
            return TITLE;
        } else if (searchResult instanceof Route) {
            return ROUTE;
        } else if (searchResult instanceof Stop) {
            return BUS_STOP;
        }
        return super.getItemViewType(position);
    }

    void setSearchResult(List<Object> busStops) {
        this.searchResults = busStops;
        notifyDataSetChanged();
    }

    public class TitleViewHolder extends RecyclerView.ViewHolder {

        private TextView title;

        TitleViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.title);
        }

        public void bind(String text) {
            title.setText(text);
        }

    }

    public class RouteViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private TextView routeId;
        private TextView routeDescription;

        RouteViewHolder(View itemView) {
            super(itemView);
            routeId = itemView.findViewById(R.id.route_id);
            routeDescription = itemView.findViewById(R.id.route_description);
        }

        public void bind(Route route) {
            routeId.setText(route.getId());
//            if (route.getOperator().equals(Operator.GO_AHEAD_DUBLIN)) {
//                routeId.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.textColorPrimaryInverse));
//                routeId.setBackgroundTintList(ColorStateList.valueOf(ContextCompat.getColor(itemView.getContext(), R.color.colorPrimary)));
//            } else {
//                routeId.setTextColor(ContextCompat.getColor(itemView.getContext(), R.color.textColorPrimary));
//                routeId.setBackgroundTintList(ColorStateList.valueOf(ContextCompat.getColor(itemView.getContext(), R.color.colorAccent)));
//            }
            routeDescription.setText(String.format(Locale.UK, "%s - %s", route.getOrigin(), route.getDestination()));
        }

        @Override
        public void onClick(View itemView) {
            int adapterPosition = getAdapterPosition();
            if (adapterPosition > -1 && adapterPosition < searchResults.size()) {
                Object searchResult = searchResults.get(adapterPosition);
                if (searchResult instanceof Route) {
                    Route route = (Route) searchResult;
                    view.launchRouteActivity(route.getId(), route.getOperator().getCode());
                }
            }
        }

    }

    public class BusStopViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private TextView stopID;
        private TextView stopName;
        private TextView routes;
        private String formattedStopId;

        BusStopViewHolder(View itemView) {
            super(itemView);
            stopID = itemView.findViewById(R.id.stop_id);
            stopName = itemView.findViewById(R.id.stop_name);
            routes = itemView.findViewById(R.id.routes);
            formattedStopId = itemView.getContext().getString(R.string.formatted_stop_id);
        }

        public void bind(Stop busStop) {
            stopID.setText(String.format(Locale.UK, formattedStopId, busStop.id()));
            stopName.setText(busStop.name());
            if (CollectionUtils.isNullOrEmpty(busStop.routes())) {
                routes.setText(StringUtils.EMPTY_STRING);
                routes.setVisibility(View.GONE);
            } else {
                String middleDot = String.format(Locale.UK, " %s ",
                        StringUtils.MIDDLE_DOT);
                routes.setText(StringUtils.join(busStop.routes(), middleDot));
                routes.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public void onClick(View itemView) {
            int adapterPosition = getAdapterPosition();
            if (adapterPosition > -1 && adapterPosition < searchResults.size()) {
                Object searchResult = searchResults.get(adapterPosition);
                if (searchResult instanceof Stop) {
                    Stop busStop = (Stop) searchResult;
                    view.launchRealTimeActivity(busStop.id());
                }
            }
        }

    }

}

File: app/src/main/kotlin/ie/dublinbuspal/view/search/SearchAdapter.kt
package ie.dublinbuspal.view.search

import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import ie.dublinbuspal.android.R
import ie.dublinbuspal.model.route.Route
import ie.dublinbuspal.model.service.Operator
import ie.dublinbuspal.model.stop.ResolvedStop
import ie.dublinbuspal.util.CollectionUtils
import ie.dublinbuspal.util.StringUtils
import kotlinx.android.synthetic.main.list_item_bus_stop.view.*
import kotlinx.android.synthetic.main.list_item_route.view.*
import kotlinx.android.synthetic.main.list_item_search_title.view.*
import java.util.*

class SearchAdapter(
        private val view: SearchView
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val TITLE = 0
    private val ROUTE = 1
    private val BUS_STOP = 2

    private var searchResults: List<Any> = emptyList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            TITLE -> {
                val titleView = LayoutInflater.from(parent.context)
                        .inflate(R.layout.list_item_search_title, parent, false)
                return TitleViewHolder(titleView)
            }
            ROUTE -> {
                val routeView = LayoutInflater.from(parent.context)
                        .inflate(R.layout.list_item_route, parent, false)
                val routeViewHolder = RouteViewHolder(routeView)
                routeView.setOnClickListener(routeViewHolder)
                return routeViewHolder
            }
            BUS_STOP -> {
                val busStopView = LayoutInflater.from(parent.context)
                        .inflate(R.layout.list_item_bus_stop, parent, false)
                val busStopViewHolder = BusStopViewHolder(busStopView)
                busStopView.setOnClickListener(busStopViewHolder)
                return busStopViewHolder
            }
            else -> throw IllegalStateException(String.format(Locale.UK,
                    "Incorrect view type: %s", viewType.toString()))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val searchResult = searchResults[position]
        if (searchResult is String && holder is TitleViewHolder) {
            holder.bind(searchResult)
        } else if (searchResult is Route && holder is RouteViewHolder) {
            holder.bind(searchResult)
        } else if (searchResult is ResolvedStop && holder is BusStopViewHolder) {
            holder.bind(searchResult)
        }
    }

    override fun getItemCount() =  searchResults.size

    override fun getItemViewType(position: Int): Int {
        val searchResult = searchResults[position]
        return when (searchResult) {
            is String -> TITLE
            is Route -> ROUTE
            is ResolvedStop -> BUS_STOP
            else -> super.getItemViewType(position)
        }
    }

    fun setSearchResult(searchResults: List<Any>) {
        this.searchResults = searchResults
        notifyDataSetChanged()
    }

    inner class TitleViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val title = itemView.title

        fun bind(text: String) {
            title.text = text
        }

    }

    inner class RouteViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        fun bind(route: Route) {
            itemView.route_id.text = route.id
            if (route.operator == Operator.GO_AHEAD_DUBLIN) {
                itemView.route_id.setTextColor(ContextCompat.getColor(itemView.context, R.color.white))
                itemView.route_id.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(itemView.context, R.color.colorPrimary))
            } else {
                itemView.route_id.setTextColor(ContextCompat.getColor(itemView.context, R.color.textColorPrimary))
                itemView.route_id.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(itemView.context, R.color.colorAccent))
            }
            itemView.route_description.text = String.format(Locale.UK, "%s - %s",
                    route.variants[0].origin, route.variants[0].destination) //TODO
        }

        override fun onClick(itemView: View) {
            val searchResult = searchResults[adapterPosition]
            if (searchResult is Route) {
                val route = searchResult
                //view.launchRouteActivity(route.id)
            }
        }

    }

    inner class BusStopViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        fun bind(busStop: ResolvedStop) {
            itemView.stop_id.text = String.format(Locale.UK, itemView.context.getString(R.string.formatted_stop_id), busStop.id())
            itemView.stop_name.text = busStop.name()
            if (CollectionUtils.isNullOrEmpty(busStop.routes())) {
                itemView.routes.visibility = View.GONE
            } else {
                itemView.routes.text = StringUtils.join(busStop.routes(), " ${StringUtils.MIDDLE_DOT} ")
            }
        }

        override fun onClick(itemView: View) {
            val searchResult = searchResults[adapterPosition]
            if (searchResult is ResolvedStop) {
                view.onBusStopClicked(searchResult.id(), searchResult.name())
            }
        }

    }

}
